subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java

```diff
@@ -117,7 +117,9 @@ public JavaCompile() {
     @Override
     @Internal("tracked via stableSources")
     public FileTree getSource() {
-        return super.getSource();
+        return compileOptions.getSourcepath() != null && !compileOptions.getSourcepath().isEmpty()
```

Comment:
I currently don't understand enough about `sourcepath` to know what the right solution is here. If we add the sourcepath as regular sources, why isn't it set to be that in the first place? And if we add the sourcepath here, then we should mark `CompileOptions.getSourcepath()` as `@Internal`, since we don't need to track it as an input any more.

Note that the way it is now, `CompileOptions.getSourcepath()` is not marked as `@Incremental`, so any change to the files there will cause a full recompilation.

Reply:
I reverted that. I figured that is not the right solution anyway. I changed it so we don't set options.sourcepath ourselves if user set them manually.

I think this is better also now it works in the same way as it did in 6.9.1

Reply:
Ah, so the problem is only if you have some modules on the classpath, right?

Reply:
Yeah it looks like it. Without module sourcepath defined by user is passed to javac.